mycpp/gc_dict.h

```diff
@@ -307,49 +284,117 @@ void Dict<K, V>::clear() {
   len_ = 0;
 }
 
-// Returns the position in the array.  Used by dict_contains(), index(),
-// get(), and set().
-//
-// For now this does a linear search.
 // TODO:
-// - hash functions, and linear probing.
-// - resizing based on load factor
-//   - which requires rehashing (re-insert all items)
 // - Special case to intern Str* when it's hashed?  How?
 //   - Should we have wrappers like:
 //   - V GetAndIntern<V>(D, &string_key)
 //   - SetAndIntern<V>(D, &string_key, value)
 //   This will enable duplicate copies of the string to be garbage collected
+//
+// NOTE: This draws a lot of inspiration from this proposal for CPython's dict.
+// The links below are useful references.
+// https://code.activestate.com/recipes/578375/
+// https://mail.python.org/pipermail/python-dev/2012-December/123028.html
 template <typename K, typename V>
-int Dict<K, V>::position_of_key(K key) {
-  for (int i = 0; i < capacity_; ++i) {
-    int special = entry_->items_[i];  // NOT an index now
-    if (special == kDeletedEntry) {
-      continue;  // keep searching
+int Dict<K, V>::hash_and_probe(K key) const {
+  if (capacity_ == 0) {
+    return kNotFound;
+  }
+
+  // Hash the key onto a slot in the index. If the first slot is occupied, probe
+  // until an empty one is found.
+  unsigned h = hash_key(key);
+  int init_bucket = h % index_len_;
+
+  // If we see a tombstone along the probing path, stash it.
+  int open_slot = -1;
+
+  for (int i = 0; i < index_len_; ++i) {
+    // Start at init_bucket and wrap araound
+    // NOTE: if division becomes a hotspot, split this into two loops.
+    int slot = (i + init_bucket) % index_len_;
+
+    int kv_index = index_->items_[slot];
+    DCHECK(kv_index < len_);
+    // Optimistically this is the common case once the table has been populated.
+    if (kv_index >= 0) {
+      unsigned h2 = hash_key(keys_->items_[kv_index]);
+      if (h == h2 && keys_equal(keys_->items_[kv_index], key)) {
+        return slot;
+      }
+    }
+
+    if (kv_index == kEmptyEntry) {
+      if (open_slot != -1) {
+        slot = open_slot;
+      }
+      // If there isn't room in the entry arrays, tell the caller to resize.
+      return len_ < capacity_ ? slot : kNotFound;
```

Comment:
hm can we `DCHECK(len_ < capacity_)` because we got kEmptyEntry, and return `slot` ?

I think these conditions are equivalent:

- there's no room in the hash table for the new entry
- every entry i `index_` is either valid, or kDeletedEntry

In that case the iteration will proceed to the end of the loop, and then `kNotFound` will be returned ?

Maybe

```
if (kv_index == kEmptyEntry) {  // we found a place to put it
  DCHECK(len_ < capacity_);  // which means this must be true
  if (open_slot == -1) {
    return slot;  // put it in this empty slot
  } else {
    return open_slot;  // put it in the FIRST place  we found kDeletedEntry ??
  }
}
```

I guess I'm also a little confused if it matters if we return the position of the FIRST kDeletedEntry or the LAST ?  are they equivalent?

Reply:
> I guess I'm also a little confused if it matters if we return the position of the FIRST kDeletedEntry on the probe path, or the LAST ? are they equivalent?

It doesn't really matter if we pick the first, last, or anything in between. As long as we wait until we finish the probe sequence to be sure we are dealing with a new key before writing to whichever one we chose we're OK and don't have to worry about disrupting the paths to existing keys. Just did a quick test locally to confirm this.

The comment at the bottom of this loop captures this pretty well, I think.